src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -1034,6 +1046,27 @@ REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalRegisterHijackCallback(_In_ PalH
     ASSERT(g_pHijackCallback == NULL);
     g_pHijackCallback = callback;
 
+#ifdef __APPLE__
```

Comment:
Is this going to work for iOS?

Reply:
It should, the new API supports iOS too.

Reply:
I don't know, but I hope that we will be able to remove the dlopen very soon, it is a matter of upgrading macOS on our CI machines.

Reply:
cc @ivanpovazan

Reply:
I am not sure I understand the changes here. What happens when the `libdispatch.dylib` is not available on the system e.g., I am currently running MacOS 14.5 and do not have the library on the given path?

Reply:
This is invoking a new API that according to Apple is present in macOS starting 14.4 and iOS/tvOS 17.4. I have originally had the commented out code there and it compiled / worked fine on my macOS device with XCode 15.4. But it turned out that our CI build machines don't have such a new XCode installed yet and there are some issues upgrading them. That's why I have modified it to this dlopen way so that we can build on older XCode and still let the runtime take advantage of this API if it is present on the target machine.
We have always had a problem in case .NET code was running on a dispatch queue thread and it was in a long running code in single method that didn't sync with GC. Then GC runtime suspension could take a long time or in case of an infinite loop never succeed. The reason was that Apple was previously blocking sending async signals to dispatch queue threads. Based on our demand, they have recently added the `dispatch_allow_send_signals` API so that we can explicitly enable that for threads running .NET code.

As for your machine not having the library, that is very strange, as you have macOS that should support this API. I have assumed that the path I've used is the path where the library is located always. Can you please see if you have it in a different location? The problem is that I was unable to dlopen it just by its name, that's why I had to use the full path.

Reply:
The file is available at `/usr/lib/system/introspection/libdispatch.dylib`:
```
ivanpovazan@EONE-MAC-ARM64 ~ $ find /usr/lib -name "libdispatch.dylib" -exec realpath {} \;
/usr/lib/system/introspection/libdispatch.dylib
ivanpovazan@EONE-MAC-ARM64 ~ $ sw_vers -productVersion 
14.5
```

Reply:
This is interesting. I also have the library in the /usr/lib/system/introspection and not in the /usr/lib/system. But when I run a simple c app that invokes that API under lldb and list the images in the process, it shows the path I have used:
```
[  8] 502762EE-7AA7-306C-9DBD-88981A86BB78 0x00000001831b2000 /usr/lib/system/libdispatch.dylib
```

The dlopen works no matter whether I use the /usr/lib/system/libdispatch.dylib or /usr/lib/system/introspection/libdispatch.dylib. However, if I load the one in the /usr/lib/system/introspection, then lldb shows many warnings like
```
objc[47199]: Class OS_dispatch_workloop is implemented in both /usr/lib/system/libdispatch.dylib (0x1eb1a3ad8) and /usr/lib/system/introspection/libdispatch.dylib (0x1001d8118). One of the two will be used. Which one is undefined.
```
So it seems that macOS has some redirection there and the path I have in the PR is the right one to use.

Reply:
dylibs don't actually exist on disk in macOS anymore, they're in a system-level dyld cache, which is why 'find' or 'ls' won't find them. dlopen knows to look in the cache though, so that's why the path just works.

Ref: https://mjtsai.com/blog/2020/06/26/reverse-engineering-macos-11-0/

One way to find out which dylib a symbol is supposed to be in is by looking in the Xcode's SDKS:

```
$ grep dispatch_allow_send_signals /Applications/Xcode_15.3.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib -R
/Applications/Xcode_15.3.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib/libSystem.B.tbd:                       _dispatch_after_f, _dispatch_allocator_layout, _dispatch_allow_send_signals,
/Applications/Xcode_15.3.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib/system/introspection/libdispatch.tbd:                       _dispatch_after_f, _dispatch_allocator_layout, _dispatch_allow_send_signals,
/Applications/Xcode_15.3.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib/system/libdispatch.tbd:                       _dispatch_after_f, _dispatch_allocator_layout, _dispatch_allow_send_signals,
/Applications/Xcode_15.3.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib/libSystem.B_asan.tbd:                       _dispatch_after_f, _dispatch_allocator_layout, _dispatch_allow_send_signals,
```

This shows that the symbol is also in libSystem.B.

One interesting thing shows up when grepping in the iOS SDK:

```
$ grep dispatch_allow_send_signals /Applications/Xcode_15.3.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/lib -R
/Applications/Xcode_15.3.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/lib/libSystem.B.tbd:                       _dispatch_after_f, _dispatch_allocator_layout, _dispatch_allow_send_signals,
/Applications/Xcode_15.3.0.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/lib/libSystem.B_asan.tbd:                       _dispatch_after_f, _dispatch_allocator_layout, _dispatch_allow_send_signals,
```

The symbol is *only* available in libSystem, not libdispatch (the same happens in tvOS too).

My recommendation would be to dlopen libSystem instead ("/usr/lib/libSystem.dylib"), hopefully that will work on all platforms.

Ideally you'd be able to remove the call to dlopen, that would reliably work on all platforms.

Reply:
> Ideally you'd be able to remove the call to dlopen, that would reliably work on all platforms.

Right, I am really waiting for the CI machines to get updated.

I'll give the libSystem a try, hopefully it would work.

Reply:
@rolfbjarne thank you for the suggestion, the libSystem works too. I'll update this PR.